File: lib/domain/cultivation_plan/gateways/plan_save_user_interaction_rule_gateway.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Gateways
      class PlanSaveUserInteractionRuleGateway
        # @return [Dtos::PlanSaveUserInteractionRuleSnapshot, nil]
        def find_by_user_id_and_source_interaction_rule_id(user_id:, source_interaction_rule_id:)
          raise NotImplementedError
        end

        # @return [Dtos::PlanSaveUserInteractionRuleSnapshot, nil]
        def find_by_user_id_and_rule_type_and_source_group_and_target_group_and_region(
          user_id:,
          rule_type:,
          source_group:,
          target_group:,
          region:
        )
          raise NotImplementedError
        end

        # @param user_id [Integer]
        # @param interaction_rule_id [Integer]
        # @param attributes [Hash]
        # @return [Dtos::PlanSaveUserInteractionRuleSnapshot]
        # @raise [Domain::Shared::Exceptions::RecordInvalid]
        def update(user_id:, interaction_rule_id:, attributes:)
          raise NotImplementedError
        end

        # @param user_id [Integer]
        # @param attributes [Hash]
        # @return [Dtos::PlanSaveUserInteractionRuleSnapshot]
        # @raise [Domain::Shared::Exceptions::RecordInvalid]
        def create(user_id:, attributes:)
          raise NotImplementedError
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/gateways/plan_save_user_interaction_rule_gateway.rs
//! Ruby: `Domain::CultivationPlan::Gateways::PlanSaveUserInteractionRuleGateway`

use crate::cultivation_plan::dtos::PlanSaveUserInteractionRuleSnapshot;
use crate::shared::attr::AttrMap;

pub trait PlanSaveUserInteractionRuleGateway: Send + Sync {
    fn find_by_user_id_and_source_interaction_rule_id(
        &self,
        user_id: i64,
        source_interaction_rule_id: i64,
    ) -> Result<Option<PlanSaveUserInteractionRuleSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn find_by_user_id_and_rule_type_and_source_group_and_target_group_and_region(
        &self,
        user_id: i64,
        rule_type: &str,
        source_group: &str,
        target_group: &str,
        region: Option<&str>,
    ) -> Result<Option<PlanSaveUserInteractionRuleSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn update(
        &self,
        user_id: i64,
        interaction_rule_id: i64,
        attributes: AttrMap,
    ) -> Result<PlanSaveUserInteractionRuleSnapshot, Box<dyn std::error::Error + Send + Sync>>;

    fn create(
        &self,
        user_id: i64,
        attributes: AttrMap,
    ) -> Result<PlanSaveUserInteractionRuleSnapshot, Box<dyn std::error::Error + Send + Sync>>;
}
